### PR TITLE
Improve testability for OpenPassClient and Token Validation

### PR DESCRIPTION
**Pull request by dcaunt**
This PR:
* Adds an async `assertThrowsError` for testing that async code throws
* Adds an HTTPStub allowing us to serve stub network responses while still using URLSession/Foundation APIs.
* Extracts validation from the network client `OpenPassClient` to its own unit
    * Adds additional coverage for ID Token validation

This unlocks being able to test the `OpenPassManager` sign-in flow by improving how we structure and test dependencies. A following PR will add those tests, and allow passing the remaining dependencies to the `OpenPassManager`.

### CODE REVIEW

File: Sources/OpenPass/IDTokenValidation.swift [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+}
+
+/// Validates the following:
+/// - Token is not expired
+/// - Token issue date is not in the past
+/// - Token signature is correct according to JWKS
+///  https://openid.net/specs/openid-connect-core-1_0.html#IDTokenValidation
+internal struct IDTokenValidator: IDTokenValidation {
+    /// Set a specific leeway window in seconds in which the Expires At ("exp") Claim will still be valid.
+    private let expiresAtLeeway: Double = 0
```
Comment (IanDBird) [inline Sources/OpenPass/IDTokenValidation.swift:41]: Not important, but just an observation... On Android, we actually expose both the `expiresAtLeeway` and `issuedAtLeeway` publicly on a per-flowclient basis. I'm trying to remember the reasoning, but I think it was just the fact that time synchronisation is tough 😅

File: Sources/OpenPass/IDTokenValidation.swift [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+    /// when the value is present
+    private let issuedAtLeeway: Double = 60
+
+    private let dateGenerator: DateGenerator
+
+    init(dateGenerator: DateGenerator = .init { Date() }) {
+        self.dateGenerator = dateGenerator
+    }
+
+    func validate(_ token: IDToken, jwks: JWKS) throws -> Bool {
```
Comment (IanDBird) [inline Sources/OpenPass/IDTokenValidation.swift:54]: Just comparing to Android's `IDTokenVerifier`, I think we're missing at least an audience (AUD) check against the client id? and potentially the token "issuer"?

File: Tests/OpenPassTests/IDTokenValidationTests.swift [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+// SOFTWARE.
+//
+
+@testable import OpenPass
+import XCTest
+
+@available(iOS 13.0, *)
+final class IDTokenValidationTests: XCTestCase {
```
Comment (IanDBird) [inline Tests/OpenPassTests/IDTokenValidationTests.swift:31]: Just checked Android's UTs, and we have two additional tests to check audience and issuer miss-matches.

**Review by IanDBird — APPROVED** [review]
Just some observations about potential validation miss-matches between iOS vs Android.

**Review by koalaparadise — COMMENTED** [review]
Looks good from a Swift point of view. I'll defer to Ian's comments though

**Pull request merged**